### PR TITLE
Subcell: add functions for the DG actions to call to support subcell

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

The DG actions `ComputeTimeDerivative` and `ApplyBoundaryCorrections` need to be aware that subcell exists. However, the DG library doesn't depend on DG-subcell because we just forward declare the functions and hide them behind an `if constexpr (using_subcell)`. This means that non-subcell executables shouldn't need to link against DG-subcell.

Note that currently the DG-FD interface are not conservative. While making them be conservative wouldn't be too terribly difficult, so far I have no evidence that this actually matters since if we are doing DG the solution is smooth anyway.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/Evolution/DgSubcell/PrepareNeighborData.hpp [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+#include "Evolution/DgSubcell/Tags/NeighborData.hpp"
+#include "NumericalAlgorithms/Spectral/Mesh.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace evolution::dg::subcell {
+/*!
+ * \brief Add local data for the relaxed discrete maximum principle
+ * troubled-cell indicator, and compute and slice data needed for the neighbor
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/PrepareNeighborData.hpp]: [super nitpicky] It may be helpful to clarify "Add local data for the neighbor's relaxed discrete maximum principle troubled-cell indicator, ..." because it can be easy to forget how this hangs together.

File: src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+namespace evolution::dg::subcell {
+/*!
+ * \brief Invoked in directions where the neighbor is doing subcell, this
+ * function computes the neighbor data on the mortar via reconstruction on
+ * nearest neighbor subcells.
+ *
+ * The max/min of the evolved variables from neighboring elements that is used
+ * for the relaxed discrete maximum principle troubled-cell indicator is
+ * combined with the data from element and stored in
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp]: "from element" -> "from the neighbor element" I assume? or local?

File: src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+ * `subcell::Tags::NeighborDataForReconstructionAndRdmpTci`. Additionally, the
+ * data needed for reconstruction is copied over into
+ * `subcell::Tags::NeighborDataForReconstructionAndRdmpTci`.
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp:36]: I would favor switching the order of sentences, I think this will make the description match the name (and therefore prior expectation) better so be less surprising: "The data needed for reconstruction... Additionally, the max/min of the evolved variables...". Could also help to add a short note of why you're handling the RDMP data at this time... it's not obvious how it's related to the reconstruction. I'm guessing you're handling it now because it's convenient but isn't needed for the reconstruction?

File: src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      neighbor_reconstructed_evolved_vars =
+          Metavariables::SubcellOptions::DgComputeSubcellNeighborPackagedData::
+              apply(*box, mortars_to_reconstruct_to);
+  ASSERT(neighbor_reconstructed_evolved_vars.size() ==
+             mortars_to_reconstruct_to.size(),
+         "Should have reconstructed "
+             << mortars_to_reconstruct_to.size() << " sides but reconstructed "
+             << neighbor_reconstructed_evolved_vars.size() << " sides.");
+  // Now copy over the packaged data _into_ the inbox so the rest of the DG
+  // algorithm is left untouched.
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp]: I have no idea what this sentence is trying to say... can you rephrase?

File: src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                neighbor_ghost_and_subcell_data[offset_for_min + var_index]);
+          }
+
+          ASSERT(subcell_neighbor_data_ptr->find(mortar_id) ==
+                     subcell_neighbor_data_ptr->end(),
+                 "The subcell neighbor data is already inserted. Direction: "
+                     << mortar_id.first
+                     << " with ElementId: " << mortar_id.second);
+          // Copy over the neighbor data for reconstruction. We need this
+          // since we might be doing a step unwind.
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp]: Can you expand on this a little? does `neighbor_ghost_and_subcell_data` get erased when unwinding? I'm not quite sure what is meant here so a few more words would help I think.

File: src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+          subcell::NeighborData neighbor_data{};
+          neighbor_data.data_for_reconstruction.insert(
+              neighbor_data.data_for_reconstruction.end(),
+              neighbor_ghost_and_subcell_data.begin(),
+              std::prev(
+                  neighbor_ghost_and_subcell_data.end(),
+                  2 * static_cast<std::ptrdiff_t>(number_of_evolved_vars)));
```
Comment (fmahebert) [inline src/Evolution/DgSubcell/NeighborReconstructedFaceSolution.hpp:131]: Help what is going on? Is this copying the first (ghost) part of `neighbor_ghost_and_subcell_data`, and omitting the subcell part? What is the "2" here? I think a comment might be useful.

**Review by fmahebert — COMMENTED** [review]
Code looks good, but I have some clarification requests on the documentation

**Comment by nilsdeppe** [comment]
I've pushed fixups that try to clarify the comments and dox. Please let me know if it is now actually clear :)

**Review by fmahebert — COMMENTED** [review]
Fixups LGTM, you can squash. Thanks for the clarifications!

**Comment by nilsdeppe** [comment]
Rebased and squashed. Thanks for the review :D

**Review by fmahebert — DISMISSED** [review]

File: src/Evolution/DgSubcell/PrepareNeighborData.hpp [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+          db::get<variables_tag>(*box),
+          db::get<evolution::dg::subcell::Tags::Inactive<variables_tag>>(*box),
+          true));
+
+  const auto& element = db::get<::domain::Tags::Element<volume_dim>>(*box);
+  DirectionMap<volume_dim, std::vector<double>>
+      all_neighbor_data_for_reconstruction{};
+
+  DirectionMap<volume_dim, bool> directions_to_slice{};
+  for (const auto& direction_neighbors : element.neighbors()) {
```
Comment (kidder) [inline src/Evolution/DgSubcell/PrepareNeighborData.hpp]: use structured binding?

File: src/Evolution/DgSubcell/PrepareNeighborData.hpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+      all_neighbor_data_for_reconstruction.at(direction) =
+          orient_variables(all_neighbor_data_for_reconstruction.at(direction),
+                           Index<volume_dim>{slice_extents}, orientation);
+    }
+
+    // Add the RDMP TCI data to what we will be sending.
+    // Note that this is added _after_ the reconstruction data has been
+    // re-oriented.
+    std::vector<double>& neighbor_scl_data =
```
Comment (kidder) [inline src/Evolution/DgSubcell/PrepareNeighborData.hpp]: scl -> subcell

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash immediately

**Comment by nilsdeppe** [comment]
I've rebased and squashed, thanks for the reviews :)

**Review by fmahebert — APPROVED** [review]

**Review by kidder — APPROVED** [review]
